modules/dnn/src/layers/gemm_layer.cpp

```diff
@@ -0,0 +1,361 @@
+        int num_inputs = static_cast<int>(inputs.size() + blobs.size());
+        CV_CheckGE(num_inputs, 2, "DNN/Gemm: Gemm takes at least two inputs");
+        CV_CheckLE(num_inputs, 3, "DNN/Gemm: Gemm takes at most three inputs");
+
+        // Check whether A and B are two dimensional
+        const auto shape_A = inputs[0];
+        const auto shape_B = const_B ? shape(blobs[0]) : inputs[1];
+        CV_CheckGE(shape_A.size(), static_cast<size_t>(2), "DNN/Gemm: Tensor A must be n-dimensional (n >= 2)");
+        CV_CheckEQ(shape_B.size(), static_cast<size_t>(2), "DNN/Gemm: Tensor B must be two dimensional");
+
+        // Check legal matrix multiplication
+        size_t dims_A = shape_A.size();
+        int ma = shape_A[dims_A - 2], na = shape_A[dims_A - 1];
+        int mb = shape_B[0], nb = shape_B[1];
+        int M = trans_a ? na : ma;
+        int N = trans_b ? mb : nb;
+        int K_a = trans_a ? ma : na;
+        int K_b = trans_b ? nb : mb;
+        CV_CheckEQ(K_a, K_b, "DNN/Gemm: Invalid dimension of dim K");
+
+        // Check whether C can be unidirectional broadcast to (M, N). Handle carefully with 1D Mat.
+        if (have_bias) {
+            const auto shape_C = const_C ? shape(blobs.back()) : inputs.back();
+
+            auto ndims_C = shape_C.size();
+            CV_CheckLE(ndims_C, static_cast<size_t>(2), "DNN/Gemm: C can only be 0d (scalar) / 1d / 2d tensor");
+
+            if (real_ndims_C == 1) { // (1,) or (N,)
+                CV_Check(shape_C[0], shape_C[0] == 1 || shape_C[0] == N, "DNN/Gemm: invalid dimension of C");
+            } else if (real_ndims_C == 2) { // (1, 1) or (1, N) or (M, 1) or (M, N)
+                // printf("shape_C=[%d, %d]\n", shape_C[0], shape_C[1]);
+                CV_Check(shape_C[0], (shape_C[0] == 1 && shape_C[1] == 1) ||
+                                     (shape_C[0] == 1 && shape_C[1] == N) ||
+                                     (shape_C[0] == M && shape_C[1] == 1) ||
+                                     (shape_C[0] == M && shape_C[1] == N),
+                                     "DNN/Gemm: C must be of shape (1, 1) or (1, N) or (M, 1) or (M, N)");
+                if (shape_C[0] == 1) {
+                    CV_Check(shape_C[1], shape_C[1] == 1 || shape_C[1] == N, "DNN/Gemm: invalid dimension of C");
+                } else if (shape_C[0] == M) {
+                    CV_Check(shape_C[1], shape_C[1] == 1 || shape_C[1] == N, "DNN/Gemm: invalid dimension of C");
+                } else {
+                    CV_Error(Error::StsBadSize, "DNN/Gemm: invalid dimension of C");
+                }
+            }
+        }
+
+        int batches = std::accumulate(shape_A.begin(), shape_A.end() - 2, 1, std::multiplies<int>());
+        MatShape shape_y{M * batches, N};
+        outputs.assign(1, shape_y);
+        return false;
+    }
+
+    // TODO: replace with cv::broadcast() once 1d mat is supported
```

Comment:
https://github.com/opencv/opencv/pull/23965 merged

Reply:
Yes, but broadcasting from a 1d tensor (which is 2d mat in opencv, for example a 1d tensor of shape [2] is loaded into a mat of shape [2, 1]; The appended 1 breaks the broadcast rule) to nd may have problems.